### ISSUE TITLE
Threading in random key to `AbstractBijection.{transform_and_log_det}` and relatives

### ISSUE THREAD

**Issue by amanjitsk** [danielward27/flowjax#224]
Hi,

How would you recommend threading in random keys for bijection methods such as `transform_and_log_det` and others? Or is that not currently possible? I figure this would facilitate approximate computation of log determinants using stochastic estimators.

**Comment by danielward27** [comment]
Maybe something like this?
```python
from typing import ClassVar

import equinox as eqx
import jax.numpy as jnp
import jax.random as jr
from jaxtyping import Array, ArrayLike, PRNGKeyArray

from flowjax.bijections import AbstractBijection


class LocWithRandomLogDet(AbstractBijection):
    """Location transform with random log det (for illustration)."""

    shape: tuple[int, ...]
    cond_shape: ClassVar[None] = None
    loc: Array
    key: PRNGKeyArray

    def __init__(
        self,
        key: PRNGKeyArray,
        loc: ArrayLike = 0,
    ):
        self.loc = jnp.asarray(loc)
        self.shape = self.loc.shape
        self.key = key

    def transform_and_log_det(self, x, condition=None):
        y = x + self.loc
        stochastic_log_det = jr.normal(self.key)
        return y, stochastic_log_det

    def inverse_and_log_det(self, y, condition=None):
        x = y - self.loc
        stochastic_log_det = jr.normal(self.key)
        return x, stochastic_log_det

    def set_key(self, key: PRNGKeyArray):
        return eqx.tree_at(lambda b: b.key, self, key)


key = jr.PRNGKey(0)
bijection = LocWithRandomLogDet(key, loc=2)

for _ in range(3):
    key, subkey = jr.split(key)
    bijection = bijection.set_key(subkey)
    y, log_det = bijection.transform_and_log_det(1)
    print(y, log_det)


# If you have these in a complex model and don't want to manually update the seeds,
# we can map over the model.

import jax

from flowjax.bijections import Chain
from flowjax.distributions import Normal, Transformed


def update_keys(model, key):
    # Generate key for each LocWithRandomLogDet with same tree structure
    treedef = jax.tree_util.tree_structure(
        model,
        is_leaf=lambda leaf: isinstance(leaf, LocWithRandomLogDet),
    )
    keys = jax.random.split(key, treedef.num_leaves)
    keys = jax.tree_util.tree_unflatten(treedef, keys)

    # Set the keys
    return jax.tree_util.tree_map(
        lambda l, key: l.set_key(key) if isinstance(l, LocWithRandomLogDet) else l,
        model,
        keys,
        is_leaf=lambda leaf: isinstance(leaf, LocWithRandomLogDet),
    )


dist = Transformed(
    base_dist=Normal(),
    bijection=Chain([LocWithRandomLogDet(key, loc=2), LocWithRandomLogDet(key, loc=2)]),
)

for _ in range(3):
    key, subkey = jr.split(key)
    dist = update_keys(dist, subkey)
    print(dist.log_prob(1))

```
To use the training loops, you'd have to modify the loss function used to make sure you update the key before each call to ``transform_and_log_det``/``inverse_and_log_det``.

**Comment by danielward27** [comment]
Looking back, you'd have to be careful with my suggestion e.g. ``log_prob`` over a batch would share a key, which I imagine wouldn't be what you want (so you'd e.g. have to vmap over keys and data to avoid that). At least in the main package, I don't think I'd add a key argument to every bijection, just because of the additional clutter and compexity - you'd have to add key splitting logic in a lot of places to support it package wide. Passing as a condition could be used, but I think the challenge again is you still need to manually add the key splitting whatever method is used, which may be easy or hard depending on the use case.

**Comment by amanjitsk** [comment]
Yep I figured! I don't think passing the key in as condition is that great of a solution either, especially in the presence of an actual condition that is used.

**Comment by danielward27** [comment]
There are definitely times where a PyTree for the condition would be nice. My initial thoughts are

- Would you expect special handling of keys, or just treat them as another array? e.g. should ``dist.log_prob``, if given a single key in condition, but a batch of ``x``, internally split the key, or just broadcast the same key? Unless I'm missing something, I think you'd need special treatment at least in the loss functions to avoid key reuse?
-  I think the important thing to know is the canonical (i.e. not batched) shape(s) for the condition, so you'd replace ``cond_shape`` with some pytree of tuples, denoting the canonical shape (which would still just be a ``tuple[int, ...]`` for simple array conditions). I do wonder though if the broadcasting for PyTree inputs could become a point of confusion.

I'd definitely take a look at a pull request for it, but without spending a bit more time thinking about the implications, it's hard to say whether I'd merge it.

**Comment by amanjitsk** [comment]
Thanks for the quick response. Here is what I think:

- As it concerns special handling of keys, no. It would be identical to the way it is now: Simply calling the vectorized version as in https://github.com/danielward27/flowjax/blob/cabcb42bcfdd12baa16c020078ec5e097efeaac0/flowjax/distributions.py#L107 This would just mean the caller should pass in a vectorized condition (along all leaves by default). Of course we could always provide more options like a PyTree of bools which specifies whether or not to vectorize the array in the corresponding leaf, but this is probably overkill for now (something like the `in_axes` specification for `eqx.filter_vmap`).

On the other hand, I do see what you are saying, but my initial thought was why use `jnp.vectorize` instead of `jax.vmap` (which only adds one dimension to the output). Of course, there might be scenarios where this would be useful (especially if we'd like conditional generation for multiple inputs across multiple conditions), but I would think that it'd be best to just align the `condition` and `x` arrays using `vmap` instead (with an additional call to reshape prior to and after calling `vmap` if needed).
 
- It probably will appear confusing at first as it concerns broadcasting of a PyTree, but I think that could be partially resolved if we replace `jnp.vectorize` with `jnp.vmap` in the `dist.log_prob`. Of course, this also means that the `VectorizedBijection` and other vectorized operations might not necessarily align with this.

But I think a simple (but tedious) resolution to all the above would be either to:
- to replace `jnp.vectorize` with `jnp.vmap`, or better yet `eqx.filter_vmap`. This way we could allow for more control over which leaves to vmap over using `in_axes` for the `condition`, while retaining all of its functionality. 
- or to instead replace the input `x` with a PyTree for which all arrays share the same leading dimensions when querying `dist.log_prob` (although this would require rethinking the entire bijection business, with `in_shape` and `out_shape` attributes, and also seems overkill).

This is just what I thought of immediately, but I'd have to think about this more. Also, with the possibility of passing in PyTree inputs or condition, we will have to think more carefully about whether `vectorize` is perhaps too unpredictable, and instead `vmap` with an `in_axes` is preferred. But I understand obviously this adds more complexity to the code. It might also make sense to look at how NNX implements its rng splitting features (obviously ignoring the fact that they allow for mutable PyTrees, which can be addressed with `eqx.tree_at`) for some ideas. Something as simple as adding randomness should be supported without the need for adding some much more code on a case-by-case basis (for the end user).

What do you think?

**Comment by amanjitsk** [comment]
Thank you! I appreciate the quick responses!

Ya, no worries this is all to brainstorm ideas anyway. I don't think there is a perfect solution, but your last piece of code seems reasonable. Its just that I thought it will scale much better if there was a way to thread keys in directly to the calls to Bijection and log_prob methods, but your last comment (I think) already suggests a fairly model-agnostic way of using keys.
